activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -125,24 +125,31 @@ def migration_context # :nodoc:
 
       def schema_migration # :nodoc:
         @schema_migration ||= begin
-                                conn = self
-                                spec_name = conn.pool.db_config.spec_name
-                                name = "#{spec_name}::SchemaMigration"
-
-                                Class.new(ActiveRecord::SchemaMigration) do
-                                  define_singleton_method(:name) { name }
-                                  define_singleton_method(:to_s) { name }
-
-                                  connection_handler.connection_pool_names.each do |pool_name|
-                                    if conn.pool == connection_handler.retrieve_connection_pool(pool_name)
-                                      self.connection_specification_name = pool_name
-                                      break
-                                    end
+                                superclass = find_connection_handler
+                                if superclass == ActiveRecord::Base
+                                  ActiveRecord::SchemaMigration
+                                else
+                                  name = "#{pool.db_config.spec_name}::SchemaMigration"
+
+                                  Class.new(superclass) do
+                                    define_singleton_method(:name) { name }
+                                    define_singleton_method(:to_s) { name }
+
+                                    include ActiveRecord::SchemaMigration::Concern
                                   end
                                 end
                               end
       end
 
+      def find_connection_handler
+        ActiveRecord::Base.connection_handlers.each do |name, handler|
+          handler.connection_pools.each do |pool|
+            return name.constantize if pool == self.pool
+          end
+        end
+        raise ConnectionNotEstablished, "Couldn't find an active connection for #{pool.db_config.spec_name.inspect}"
+      end
```

Comment:
This change might seem a bit weird, and I agree, however I'm afraid it's a lesser evil until we get further.

Later on we should have an object tree that represent the different databases, so it will be easy to simply create a sublcass, and give it a reference to the schema it maps to.

However right now, the topology is still very much held by the model classes, so only proper way to access the `animals` connection, is to inherit from `AnimalBase`, hence this change. The `constantize` is just an unfortunate consequence of this.

But again, I'm confident we'll be able to clean that up down the road, so I hope it's not too much of a deal breaker.